### PR TITLE
Better "why is this nil?" error message

### PR DESCRIPTION
**Pull request by jez**
### Motivation

**Before**
Screen Shot 2022-01-31 at 12 04 17 PM

**After**

Screen Shot 2022-01-31 at 12 56 51 PM

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jez** [comment]
Some followup issues:

- Can we handle `.is_a?(...) && ...` and `!(...).nil? && ...`?
- Can we handle `(...).nil? || ...`?
- This doesn't generalize to arbitrary `if` conditions—is there another approach that might work better for more cases for free?

File: ast/desugar/Desugar.cc [hunk 1]
```diff
@@ -1059,9 +1059,32 @@ ExpressionPtr node2TreeImpl(DesugarContext dctx, unique_ptr<parser::Node> what)
...
+                    auto andAndTemp = dctx.freshNameUnique(core::Names::andAnd());
+
+                    auto *lhsSend = ast::cast_tree<ast::Send>(lhs);
+                    auto *rhsSend = ast::cast_tree<ast::Send>(rhs);
+
+                    ExpressionPtr thenp;
+                    if (lhsSend != nullptr && rhsSend != nullptr) {
+                        auto lhsSource = core::Loc(dctx.ctx.file, lhsSend->loc).source(dctx.ctx);
+                        auto rhsRecvSource = core::Loc(dctx.ctx.file, rhsSend->recv.loc()).source(dctx.ctx);
+                        if (lhsSource.has_value() && lhsSource == rhsRecvSource) {
```
Comment (aprocter) [inline ast/desugar/Desugar.cc:1071]: So it looks like we're only inserting `checkAndAnd`s when the lhs and the rhs receiver are lexically identical, right? So for example, `x.foo && x . foo . bar` wouldn't get one. (I think that's probably reasonable, if I'm understanding correctly.)

File: core/types/calls.cc [hunk 2]
```diff
@@ -2451,6 +2451,140 @@ class Magic_selfNew : public IntrinsicMethod {
     }
 } Magic_selfNew;
 
+class Magic_checkAndAnd : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        // args[0] is recv
+        // args[1] is the && tmp var
+        // args[2] is the method name
+        // args[3...] are the args
```
Comment (aprocter) [inline core/types/calls.cc:2460]: A high-level comment on what we expect the desugarer to produce as it relates to the original `lhs && rhs.foo`, and how we consume that info, might be nice here.

File: core/types/calls.cc [hunk 3]
```diff
@@ -2451,6 +2451,140 @@ class Magic_selfNew : public IntrinsicMethod {
...
+            selfTy.type,
+            args.block,
+            args.originForUninitialized,
+            args.isPrivateOk,
+            args.suppressErrors,
+        };
+        auto dispatched = selfTy.type.dispatchCall(gs, innerArgs);
+
+        auto multipleComponents = dispatched.secondary != nullptr;
+        if (multipleComponents) {
```
Comment (aprocter) [inline core/types/calls.cc:2506]: Why can we bail out on `!multipleComponents`? Is it because nilable is a `T.any`?

File: core/types/calls.cc [hunk 4]
```diff
@@ -2451,6 +2451,140 @@ class Magic_selfNew : public IntrinsicMethod {
...
+                    foundErrorOnRetry |= !it->main.errors.empty();
+                }
+
+                if (!foundErrorOnRetry) {
+                    for (auto it = &dispatched; it != nullptr; it = it->secondary.get()) {
+                        for (auto &err : it->main.errors) {
+                            if (err->what == core::errors::Infer::UnknownMethod && it->main.receiver.isNilClass()) {
+                                if (auto newErr = gs.beginError(err->loc, core::errors::Infer::CallAfterAndAnd)) {
+                                    newErr.setHeader(
+                                        "Call to method `{}` after `{}` assumes result type doesn't change",
```
Comment (aprocter) [inline core/types/calls.cc:2543]: I wonder if this phrasing will be a bit confusing, since "it's still nilable" doesn't exactly jump out here. Maybe something like "Receiver of call to method foo is still nilable even after &&"?

File: test/cli/autocorrect_and_and/autocorrect_and_and.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+class A
+  extend T::Sig
+
+  sig {returns(T.nilable(Integer))}
+  def foo; nil; end
+
+  def bar
+    a = A.new
+
+    if a.foo && a.foo.even?
```
Comment (aprocter) [inline test/cli/autocorrect_and_and/autocorrect_and_and.rb:12]: Mildly interesting corner case you might test (maybe this isn't the right spot for it): if there's whitespace before `even?`, e.g. `a.foo.<NL>even?` (I think the autocorrect should not fire since you're looking for a period immediately preceding.)

**Review by aprocter — APPROVED** [review]
Good stuff!

File: ast/desugar/Desugar.cc (hunk 1 above)
Comment (jez) [reply ast/desugar/Desugar.cc:1071]: Yeah, that's right. it's not great

File: core/types/calls.cc (hunk 2 above)
Comment (jez) [reply core/types/calls.cc:2460]: hm, isn't that what this comment is? i'm not sure what more you'd like

File: core/types/calls.cc (hunk 3 above)
Comment (jez) [reply core/types/calls.cc:2506]: each component in the DispatchResult corresponds to one branch of the `T.any`

The way that dispatchCall works is that it descends down the structure of the `T.any`, calls `dispatchCall` on each component, and then builds back up a `DispatchResult` that records both sides of the dispatch. Check out the code for `OrType::dispatchCall`

the error message doesn't make sense if there is less than one component—if there is only one component, then we should leave any original error message(s) untouched.

**Comment by jez-stripe** [comment]
test

File: core/types/calls.cc (hunk 2 above)
Comment (jez-stripe) [reply core/types/calls.cc:2460]: Ah gotcha. I would have assumed that to be implied from the fact that this is an intrinsic for a `<Magic>` method

Comment (aprocter) [reply core/types/calls.cc:2460]: I guess my basic point is that the purpose of the `<Magic>` method itself isn't fully clear unless you go searching for the code that emits it, so documenting that would be good. (That applies to a lot of the `<Magic>` stuff, though!)

Comment (jez) [reply core/types/calls.cc:2465]: (i added a comment)

File: test/cli/autocorrect_and_and/autocorrect_and_and.rb (hunk 5 above)
Comment (jez) [reply test/cli/autocorrect_and_and/autocorrect_and_and.rb:12]: great test!

File: ast/desugar/Desugar.cc (hunk 1 above)
Comment (jez) [reply ast/desugar/Desugar.cc:1071]: i added a test for this.

File: core/types/calls.cc (hunk 2 above)
Comment (aprocter) [reply core/types/calls.cc:2465]: Thanks! That's exactly what I was looking for.
